File: baseSrc/andriodGL/app/src/main/java/z3d/display/role/SceneChar.java
package z3d.display.role;

import z3d.scene.Scene3D;
import z3d.skill.Skill;
import z3d.skill.SkillManager;

public class SceneChar extends Display3dMovie {

    public static String WEAPON_PART="weapon";
    public static String WEAPON_DEFAULT_SLOT="w_01";
    public static String MOUNT_SLOT="mount_01";
    public static String WING_SLOT="wing_01";
    public static String SEL_PART="select";
    public static String NONE_SLOT="none";

    public static String CharAction_stand         ="stand";
    public static String CharAction_walk          ="walk";
    public static String CharAction_jump          ="jump";
    public static String CharAction_death         ="death";
    public static String CharAction_injured       ="injured";
    public static String CharAction_stand_mount      ="stand_mount";
    public static String CharAction_walk_mount      ="walk_mount";


    public static String CharAction_stand_mount_01         ="stand_mount_01";
    public static String CharAction_walk_mount_01          ="walk_mount_01";

    public MountChar mountChar;
    public SceneChar(Scene3D val) {
        super(val);
    }
    public void setMountById(String mountName) {

        if(this.mountChar==null){
            this.mountChar=new MountChar(this.scene3d);
            this.scene3d.addMovieDisplay(this.mountChar);
            this.setBind(this.mountChar,SceneChar.MOUNT_SLOT);
        }
        this.mountChar.setRoleUrl("role/"+mountName+".txt");
        this.changeMountAction();
        this.refrishmountPos();
    }

    @Override
    public void upData() {
        refrishmountPos();
        super.upData();
    }
    public boolean play(String action,int complete,boolean needFollow){
        boolean  temp=super.play(action,complete,needFollow);
        if(this.mountChar!=null){
            changeMountAction();
        }
        return temp;
    }
    private void refrishmountPos() {
        if(this.mountChar!=null){
            this.mountChar.x=this.x;
            this.mountChar.y=this.y;
            this.mountChar.z=this.y;
            this.mountChar.rotationY=this.rotationY;
        }
    }
    private void changeMountAction() {
        String action=this.curentAction;
        if(this.mountChar!=null){
            if( action.equals(CharAction_stand)||action.equals(CharAction_stand_mount_01)){
                this.curentAction=CharAction_stand_mount;
                this.mountChar.curentAction=CharAction_stand;
            }
        else if(action.equals(CharAction_walk)||action.equals(CharAction_walk_mount_01)){
                this.curentAction=CharAction_walk_mount;
                this.mountChar.curentAction=CharAction_walk;
            }else{
                this.mountChar.curentAction=CharAction_stand;
            }


        }
    }

    public void playSkill(Skill skill) {
        SkillManager.getInstance().playSkill(skill);
    }
}

File: listios/StorageStarted/pan3d/display/role/SceneChar.m
//
//  SceneChar.m
//  iosgl
//
//  Created by zhao on 18/4/2020.
//  Copyright © 2020 zhao. All rights reserved.
//

#import "SceneChar.h"
#import "Skill.h"
#import "Scene3D.h"
#import "SkillManager.h"
@interface SceneChar()
 
@property(nonatomic,strong)Skill* skillVo;
@end
@implementation SceneChar
+(NSString*)WEAPON_PART;
{
    return @"weapon";
}
+(NSString*)WEAPON_DEFAULT_SLOT;
{
    return @"w_01";
}
+(NSString*)MOUNT_SLOT;
{
    return @"mount_01";
}
+(NSString*)WING_SLOT;
{
    return @"wing_01";
}
+(NSString*)SEL_PART;
{
    return @"select";
}
+(NSString*)NONE_SLOT;
{
    return @"none";
}


+(NSString*)CharAction_stand;
{
    return @"stand";
}
+(NSString*)CharAction_walk;
{
    return @"walk";
}
+(NSString*)CharAction_jump;
{
    return @"jump";
}
+(NSString*)CharAction_death;
{
    return @"death";
}
+(NSString*)CharAction_injured;
{
    return @"injured";
}
+(NSString*)CharAction_stand_mount;
{
    return @"walk_mount";
}

+(NSString*)CharAction_stand_mount_01;
{
    return @"stand_mount_01";
}
+(NSString*)CharAction_walk_mount_01;
{
    return @"walk_mount_01";
}
 
 
-(void)setWeaponByAvatar:(int)avatar;
{
    
}
-(BOOL)play:(NSString*)action completeState:(int)completeState needFollow:(BOOL)needFollow;
{
    BOOL temp= [super play:action completeState:completeState needFollow:needFollow];
    [self changeMountAction];
    return  temp;
}
-(void)changeMountAction;
{
    SceneChar* this=self;
    NSString* action=this.curentAction;
    if(self.mountChar){
        action=action?action:this.defaultAction;
        if([action isEqualToString:CharAction_stand]||[action isEqualToString:CharAction_stand_mount]){
            this.curentAction=CharAction_stand_mount;
            self.mountChar.curentAction=CharAction_stand_mount;
//            CharAction_stand_mount
        }
        else if([action isEqualToString:CharAction_walk]||[action isEqualToString:CharAction_walk_mount]){
            this.curentAction=CharAction_walk_mount;
            self.mountChar.curentAction=CharAction_walk;
        }else{
            self.mountChar.curentAction=CharAction_walk_mount;
        }
    }
}
-(void)setMountById:(NSString*)mountName;
{
    SceneChar* this=self;
    if(!this.mountChar){
        this.mountChar=[[MountChar alloc]init:self.scene3D];
        [this.scene3D addMovieDisplay:this.mountChar];
        [this setBind:this.mountChar bindSocket:SceneChar.MOUNT_SLOT];
 
    }
    [this.mountChar setRoleUrl: getRoleUrl(mountName)];
    [self changeMountAction];
    [self refrishmountPos];
 
}
/*
 刷新坐骑位置
 */
- (void)refrishmountPos;
{
    SceneChar* this=self;
    if(this.mountChar){
        this.mountChar.x=this.x;
        this.mountChar.y=this.y;
        this.mountChar.z=this.y;
        this.mountChar.rotationY=this.rotationY;
    }
}
-(void)setX:(float)value; {
    [super setX:value];
    [self refrishmountPos];
}
-(void)setY:(float)value;{
    [super setY:value];
    [self refrishmountPos];
}
-(void)setZ:(float)value;{
    [super setZ:value];
    [self refrishmountPos];
}
 
 
- (void)upFrame;
{
     [super upFrame];
}
@end

